server/data/data-jmap/src/main/java/org/apache/james/jmap/api/change/MailboxChange.java

```diff
@@ -129,198 +123,170 @@ public MailboxChange build() {
+            MailboxChange ownerChange = MailboxChange.builder()
+                .accountId(AccountId.fromUsername(mailboxACLUpdated.getUsername()))
+                .state(stateFactory.generate())
+                .date(now)
+                .isCountChange(false)
+                .updated(ImmutableList.of(mailboxACLUpdated.getMailboxId()))
+                .build();
+
+            Stream<MailboxChange> shareeChanges = sharees.stream()
+                .map(name -> MailboxChange.builder()
+                    .accountId(AccountId.fromString(name))
                     .state(stateFactory.generate())
                     .date(now)
                     .isCountChange(false)
                     .updated(ImmutableList.of(mailboxACLUpdated.getMailboxId()))
-                    .build();
-
-                Stream<MailboxChange> shareeChanges = getSharees(mailboxACLUpdated.getMailboxId(), mailboxACLUpdated.getUsername(), mailboxManager)
-                    .map(name -> MailboxChange.builder()
-                        .accountId(AccountId.fromString(name))
-                        .state(stateFactory.generate())
-                        .date(now)
-                        .isCountChange(false)
-                        .updated(ImmutableList.of(mailboxACLUpdated.getMailboxId()))
-                        .delegated()
-                        .build());
+                    .delegated()
+                    .build());
 
-                return Stream.concat(Stream.of(ownerChange), shareeChanges)
-                    .collect(Guavate.toImmutableList());
-            }
-            if (event instanceof MailboxDeletion) {
-                MailboxDeletion mailboxDeletion = (MailboxDeletion) event;
+            return Stream.concat(Stream.of(ownerChange), shareeChanges)
+                .collect(Guavate.toImmutableList());
+        }
 
-                MailboxChange ownerChange = MailboxChange.builder()
-                    .accountId(AccountId.fromUsername(mailboxDeletion.getUsername()))
+        public List<JmapChange> fromMailboxDeletion(MailboxDeletion mailboxDeletion, ZonedDateTime now) {
+            MailboxChange ownerChange = MailboxChange.builder()
+                .accountId(AccountId.fromUsername(mailboxDeletion.getUsername()))
+                .state(stateFactory.generate())
+                .date(now)
+                .isCountChange(false)
+                .destroyed(ImmutableList.of(mailboxDeletion.getMailboxId()))
+                .build();
+
+            Stream<MailboxChange> shareeChanges = mailboxDeletion.getMailboxACL()
+                .getEntries().keySet()
+                .stream()
+                .filter(rfc4314Rights -> !rfc4314Rights.isNegative())
+                .filter(rfc4314Rights -> rfc4314Rights.getNameType().equals(MailboxACL.NameType.user))
+                .map(MailboxACL.EntryKey::getName)
+                .map(name -> MailboxChange.builder()
```

Comment:
Shouldn't the sharee be passed as a method argument?

Reply:
Not necessary. The ACLs are already defined in the MailboxDeletion event, no need to check them again?